go/vt/vtctl/workflow/materializer.go

```diff
@@ -605,3 +605,74 @@ func (mz *materializer) checkTZConversion(ctx context.Context, tz string) error
 	})
 	return err
 }
+
+func (mz *materializer) filterSourceShards(targetShard *topo.ShardInfo) []*topo.ShardInfo {
+	// Don't create streams from sources which won't contain data for the
+	// target shard. Only do it for MoveTables where the source and target
+	// keyspaces are both sharded and have the same primary vindexes.
+	if mz.samePrimaryVindexes && mz.ms.MaterializationIntent == vtctldatapb.MaterializationIntent_MOVETABLES {
+		return useIntersectingSourceShards(targetShard, mz.sourceShards)
+	}
+	return useAllSourceShards(targetShard, mz.sourceShards)
+}
+
+// samePrimaryVindexes returns true if, for all tables defined in the provided
+// materialize settings, the provided source and target vschemas have the same
+// primary vindexes.
+func samePrimaryVindexes(ms *vtctldatapb.MaterializeSettings, source, target *vschema.Keyspace) bool {
+	// For keyspaces that are not unsharded, there are no primary vindexes to
+	// compare.
+	if !source.Sharded || !target.Sharded {
+		return false
+	}
+
+	// For source and target keyspaces that are sharded, we can optimize source
+	// shard selection if source and target tables' primary vindexes are equal.
+	//
+	// To determine this, iterate over all target tables, looking for primary
+	// vindexes that differ from the corresponding source table.
+	for _, ts := range ms.TableSettings {
+		st, sok := source.Tables[ts.TargetTable]
+		tt, tok := target.Tables[ts.TargetTable]
+		if !sok || !tok {
+			// Expected vschema to have a table definition, but did not.
+			return false
+		}
+		if st.Type != "" || tt.Type != "" {
+			// Expected tables to be type "" (regular, sharded), but were not.
+			return false
+		}
+		if len(st.ColumnVindexes) == 0 || len(tt.ColumnVindexes) == 0 {
+			// Expected column to have a primary column vindex, but none
+			// was found.
```

Comment:
I like your suggestion, but would it make more sense to use `&&` instead of `||`? If the source specifies column vindexes, but the target does not, then they are not identical.